File: arduinobot_task_planner/arduinobot_task_planner/collision_monitor.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from std_msgs.msg import Bool
import math


class CollisionMonitor(Node):
    def __init__(self):
        super().__init__('collision_monitor')

        # Subscriber to joint states
        self.joint_state_subscriber = self.create_subscription(
            JointState,
            '/joint_states',
            self.joint_state_callback,
            10
        )

        # Publisher for danger collision warnings
        self.danger_publisher = self.create_publisher(
            Bool,
            '/collision_warning',
            10
        )

        # Obstacle position - Must match the one defined in scene_manager.py
        self.obstacle_position = {
            'x': 0.15,
            'y': 0.15,
            'z': 0.15
        }

        # Distance threshold in meters - A warning is triggered below this distance
        self.DANGER_THRESHOLD = 0.2  # previous value was 0.15

        # Robot link lengths in meters - Adjust these values to match your physical robot
        self.L1 = 0.12  # link_1 length
        self.L2 = 0.12  # link_2 length
        self.L3 = 0.08  # link_3 length

        self.get_logger().info(
            f'The monitoring collision node started — threshold: {self.DANGER_THRESHOLD}m'
        )

    def forward_kinematics(self, joint1, joint2, joint3):
        """
        Simplified forward kinematics used to estimate
        the end-effector position (in x-y plane).
        """
        x = (self.L1 * math.cos(joint1) +
             self.L2 * math.cos(joint1 + joint2) +
             self.L3 * math.cos(joint1 + joint2 + joint3))

        y = (self.L1 * math.sin(joint1) +
             self.L2 * math.sin(joint1 + joint2) +
             self.L3 * math.sin(joint1 + joint2 + joint3))

        z = 0.0  # planar simplification

        return x, y, z

    def distance_to_obstacle(self, x, y, z):
        """Compute the Euclidean distance from the obstacle."""
        dx = x - self.obstacle_position['x']
        dy = y - self.obstacle_position['y']
        dz = z - self.obstacle_position['z']

        return math.sqrt(dx**2 + dy**2 + dz**2)

    def joint_state_callback(self, msg):
        # Extract joint positions
        try:
            joint1_index = msg.name.index('joint_1')
            joint2_index = msg.name.index('joint_2')
            joint3_index = msg.name.index('joint_3')
        except ValueError:
            return

        joint1 = msg.position[joint1_index]
        joint2 = msg.position[joint2_index]
        joint3 = msg.position[joint3_index]

        # Compute end-effector position
        x, y, z = self.forward_kinematics(joint1, joint2, joint3)

        # Compute distance to the obstacle
        distance = self.distance_to_obstacle(x, y, z)

        # Publish warning if robot distance to obstacle is below the threshold
        warning = Bool()
        warning.data = distance < self.DANGER_THRESHOLD
        self.danger_publisher.publish(warning)

        if warning.data:
            self.get_logger().warn(
                f'WARNING — obstacle distance: {distance:.3f}m '
                f'(threshold: {self.DANGER_THRESHOLD}m)'
            )
        else:
            self.get_logger().info(
                f'Obstacle distance: {distance:.3f}m — SAFE DISTANCE',
                throttle_duration_sec=2.0
            )


def main(args=None):
    rclpy.init(args=args)
    node = CollisionMonitor()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_task_planner_cpp/src/collision_monitor.cpp
#include <memory>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "std_msgs/msg/bool.hpp"

class CollisionMonitor : public rclcpp::Node
{
public:
    CollisionMonitor()
        : Node("collision_monitor")
    {
        // Subscriber
        joint_state_subscriber_ = this->create_subscription<sensor_msgs::msg::JointState>(
            "/joint_states",
            10,
            std::bind(&CollisionMonitor::jointStateCallback, this, std::placeholders::_1));

        // Publisher
        danger_publisher_ = this->create_publisher<std_msgs::msg::Bool>(
            "/collision_warning",
            10);

        // Obstacle position
        obstacle_x_ = 0.15;
        obstacle_y_ = 0.15;
        obstacle_z_ = 0.15;

        // Distance threshold
        DANGER_THRESHOLD_ = 0.2;

        // Robot link lengths
        L1_ = 0.12;
        L2_ = 0.12;
        L3_ = 0.08;

        RCLCPP_INFO(
            this->get_logger(),
            "Collision monitor started - threshold: %.2f m",
            DANGER_THRESHOLD_);
    }

private:

    void forwardKinematics(
        double joint1,
        double joint2,
        double joint3,
        double &x,
        double &y,
        double &z)
    {
        x = L1_ * std::cos(joint1)
          + L2_ * std::cos(joint1 + joint2)
          + L3_ * std::cos(joint1 + joint2 + joint3);

        y = L1_ * std::sin(joint1)
          + L2_ * std::sin(joint1 + joint2)
          + L3_ * std::sin(joint1 + joint2 + joint3);

        z = 0.0;
    }

    double distanceToObstacle(double x, double y, double z)
    {
        double dx = x - obstacle_x_;
        double dy = y - obstacle_y_;
        double dz = z - obstacle_z_;

        return std::sqrt(dx * dx + dy * dy + dz * dz);
    }

    void jointStateCallback(const sensor_msgs::msg::JointState::SharedPtr msg)
    {
        auto findJointIndex = [&](const std::string &name) -> int
        {
            auto it = std::find(msg->name.begin(), msg->name.end(), name);

            if (it == msg->name.end())
                return -1;

            return std::distance(msg->name.begin(), it);
        };

        int joint1_index = findJointIndex("joint_1");
        int joint2_index = findJointIndex("joint_2");
        int joint3_index = findJointIndex("joint_3");

        if (joint1_index < 0 || joint2_index < 0 || joint3_index < 0)
            return;

        double joint1 = msg->position[joint1_index];
        double joint2 = msg->position[joint2_index];
        double joint3 = msg->position[joint3_index];

        double x, y, z;
        forwardKinematics(joint1, joint2, joint3, x, y, z);

        double distance = distanceToObstacle(x, y, z);

        std_msgs::msg::Bool warning;
        warning.data = (distance < DANGER_THRESHOLD_);

        danger_publisher_->publish(warning);

        if (warning.data)
        {
            RCLCPP_WARN(
                this->get_logger(),
                "WARNING - obstacle distance: %.3f m (threshold: %.3f m)",
                distance,
                DANGER_THRESHOLD_);
        }
        else
        {
            RCLCPP_INFO_THROTTLE(
                this->get_logger(),
                *this->get_clock(),
                2000,
                "Obstacle distance: %.3f m - SAFE DISTANCE",
                distance);
        }
    }

    // Subscribers
    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_state_subscriber_;

    // Publishers
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr danger_publisher_;

    // Obstacle position
    double obstacle_x_;
    double obstacle_y_;
    double obstacle_z_;

    // Robot parameters
    double L1_;
    double L2_;
    double L3_;

    // Threshold
    double DANGER_THRESHOLD_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<CollisionMonitor>();

    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;
}